File: algorithms/2_medium/M/02918_minimum_equal_sum_of_two_arrays_after_replacing_zeros/solution.py
class Solution:
    def calc_sum_and_zeros(self, nums: List[int]) -> (int, int):
        total_sum = 0
        zeros = 0

        for num in nums:
            total_sum += num

            if num == 0:
                zeros += 1

        return (total_sum, zeros)

    def minSum(self, nums1: List[int], nums2: List[int]) -> int:
        (sum1, zeros1) = self.calc_sum_and_zeros(nums1)
        (sum2, zeros2) = self.calc_sum_and_zeros(nums2)
        min_valid_sum1 = sum1 + zeros1
        min_valid_sum2 = sum2 + zeros2

        if (zeros1 == 0 and sum1 < min_valid_sum2) or (zeros2 == 0 and min_valid_sum1 > sum2):
            return -1

        return max(min_valid_sum1, min_valid_sum2)

File: algorithms/2_medium/M/02918_minimum_equal_sum_of_two_arrays_after_replacing_zeros/solution.rs
impl Solution {
    #[inline(always)]
    fn calc_sum_and_zeros(nums: &Vec<i32>) -> (i64, i64) {
        let mut sum = 0_i64;
        let mut zeros = 0_i64;

        for &num in nums.iter() {
            sum += num as i64;

            if num == 0 {
                zeros += 1;
            }
        }

        (sum, zeros)
    }

    pub fn min_sum(nums1: Vec<i32>, nums2: Vec<i32>) -> i64 {
        let (sum1, zeros1) = Self::calc_sum_and_zeros(&nums1);
        let (sum2, zeros2) = Self::calc_sum_and_zeros(&nums2);
        let min_valid_sum1 = sum1 + zeros1;
        let min_valid_sum2 = sum2 + zeros2;

        if (zeros1 == 0 && sum1 < min_valid_sum2) || (zeros2 == 0 && min_valid_sum1 > sum2) {
            return -1;
        }

        min_valid_sum1.max(min_valid_sum2)
    }
}
